libs/@local/hashql/eval/src/graph/read/filter.rs

```diff
@@ -0,0 +1,158 @@
+use hashql_hir::node::{
+    Node,
+    kind::NodeKind,
+    r#let::Let,
+    operation::{
+        BinaryOperation, Operation, OperationKind, TypeOperation,
+        binary::BinOpKind,
+        r#type::{TypeAssertion, TypeOperationKind},
+    },
+    variable::{LocalVariable, Variable, VariableKind},
+};
+
+use super::{
+    CompilationError, FilterCompilerContext, GraphReadCompiler,
+    error::qualified_variable_unsupported, path::CompleteQueryPath,
+};
+
+impl<'env, 'heap: 'env> GraphReadCompiler<'env, 'heap> {
+    #[expect(clippy::too_many_lines, reason = "match statement")]
+    pub(super) fn compile_filter<R>(
+        &mut self,
+        context: FilterCompilerContext<'heap>,
+        node: &'heap Node<'heap>,
+    ) -> Result<Filter<'heap, R>, CompilationError>
+    where
+        R: QueryRecord<QueryPath<'heap>: CompleteQueryPath<'heap, PartialQueryPath: Debug>>,
+    {
+        match node.kind {
+            NodeKind::Variable(Variable {
+                span: _,
+                kind:
+                    VariableKind::Local(LocalVariable {
+                        span: _,
+                        name,
+                        arguments: _,
+                    }),
+            }) => {
+                debug_assert_ne!(
+                    name.value, context.param_name,
+                    "typecheck should have caught this, cannot just return the entity itself."
+                );
+
+                let value = self.locals[&name.value];
+                self.compile_filter(context, value)
+            }
+            NodeKind::Variable(Variable {
+                span: _,
+                kind: VariableKind::Qualified(qualified),
+            }) => {
+                self.diagnostics
+                    .push(qualified_variable_unsupported(context, qualified));
+
+                Err(CompilationError)
+            }
+            NodeKind::Let(Let {
+                span: _,
+                name,
+                value,
+                body,
+            }) => {
+                self.locals.insert(name.value, value);
+                let filter = self.compile_filter(context, body);
+                self.locals.remove(&name.value);
+
+                filter
+            }
+
+            NodeKind::Operation(Operation {
+                span: _,
+                kind:
+                    OperationKind::Binary(BinaryOperation {
+                        span: _,
+                        op,
+                        left,
+                        right,
+                    }),
+            }) => {
+                let func = match op.kind {
+                    BinOpKind::And => {
+                        let (left, right) = (
+                            self.compile_filter(context, left),
+                            self.compile_filter(context, right),
+                        );
+
+                        return Ok(Filter::All(vec![left?, right?]));
+                    }
```

Comment:
I potentially see an issue here (which I expect can be solved later). If we have a filter with 100k entity Ids all OR-ed together, e.g. `entity_id == X OR entity_id == Y ...`, this nests _very_deep. This is one of the reason why we have `All` as `Filter` and not `And` (and `Any` instead of `Or`). Any plans for optimizing this as a special node?

Reply:
I expected that the `SelectCompiler` optimizes these during the compliation into `Conditions`. I can instead take a look at doing that flattening directly inside this stage instead.

Reply:
That's a very low-hanging fruit, I think it makes sense. Also to move more sophisticated optimizations to the DA. Thank you!